lib/sharing/gists.dart

```diff
@@ -170,92 +174,166 @@ $styleRef$dartRef  </head>
-        throw const GistLoaderException(GistLoaderFailureType.unknown);
-      }
+    // TODO(redbrogdon) This should be removed once Flutter compilation in
+    // dart-services is unforked.
+    //
+    // Remove everything up to and including main(), and replace it with a valid
+    // set of flutter_web imports and a new main() that waits for platform
+    // readiness.
+    final spliceIndex = response.body.indexOf(_sampleMain) + _sampleMain.length;
+    final modifiedCode = '$_dartPadMain${response.body.substring(spliceIndex)}';
+
+    final mainFile = GistFile(
+      name: 'main.dart',
+      content: modifiedCode,
+    );
+
+    final gist = Gist(files: [mainFile]);
+
+    if (afterLoadHook != null) {
+      afterLoadHook(gist);
     }
+
+    return gist;
+  }
+
+  String _extractContents(String githubResponse) {
+    // GitHub's API returns file contents as the "contents" field in a JSON
+    // object. The field's value is in base64 encoding, but with line ending
+    // characters ('\n') included.
+    final contentJson = json.decode(githubResponse);
+    final encodedContentStr =
+        contentJson['content'].toString().replaceAll('\n', '');
+    return utf8.decode(base64.decode(encodedContentStr));
   }
 
-  /// Create a new gist and return the newly created Gist.
-  Future<Gist> createAnon(Gist gist) {
-    // POST /gists
-    if (beforeSaveHook != null) {
-      gist = gist.clone();
-      beforeSaveHook(gist);
+  String _buildContentsUrl(String owner, String repo, String path,
+      [String ref]) {
+    return '$_repoContentsApiUrl/$owner/$repo/contents/$path'
+        '${(ref != null && ref.isNotEmpty) ? '?ref=$ref' : ''}';
+  }
+
+  Future<Gist> loadGistFromRepo({
+    String owner,
+    String repo,
+    String path,
+    String ref,
+  }) async {
+    // Download and parse the exercise's `dartpad-metadata.json` file.
+    final metadataUrl =
+        _buildContentsUrl(owner, repo, '$path/$_metadataFilename', ref);
+    final metadataResponse = await _client.get(metadataUrl);
+
+    if (metadataResponse.statusCode == 404) {
+      throw GistLoaderException(GistLoaderFailureType.contentNotFound);
+    } else if (metadataResponse.statusCode == 403) {
+      throw GistLoaderException(GistLoaderFailureType.rateLimitExceeded);
+    } else if (metadataResponse.statusCode != 200) {
+      throw GistLoaderException(GistLoaderFailureType.unknown);
     }
 
-    return HttpRequest.request(_githubApiUrl,
-            method: 'POST', sendData: gist.toJson())
-        .then((HttpRequest request) {
-      Gist gist = Gist.fromMap(json.decode(request.responseText));
-      if (afterLoadHook != null) {
-        afterLoadHook(gist);
+    final metadataContent = _extractContents(metadataResponse.body);
+
+    ExerciseMetadata metadata;
+
+    try {
+      metadata = ExerciseMetadata.fromJson(json.decode(metadataContent));
+    } on MetadataException catch (ex) {
+      print('Issue loading & parsing metadata: ${ex.message}');
```

Comment:
do we need the print here? Or should printing be done by the catcher?

Reply:
We need it somewhere for folks to diagnose issues with their metadata, but it doesn't need to be here. I'll add a message field to the exception and move the print up a level to the embed.